dev-overhaul.md

```diff
@@ -0,0 +1,154 @@
+
+We will also upgrade our communication processes to keep the Salt community more aware of the release timelines. We will share progress with the [#salt channel on IRC on Freenode](http://webchat.freenode.net/?channels=salt&uio=Mj10cnVlJjk9dHJ1ZSYxMD10cnVl83), the #release channel in the [SaltStack Community Slack](https://saltstackcommunity.herokuapp.com/), and the [Salt Users mailing list](https://groups.google.com/forum/#!forum/salt-users).
+
+## Versioning (Naming)
+
+With the `neon` release, to indicate the new change in release process, Salt will change to a new, non-date based version schema beginning at 3000. The version will be MAJOR.PATCH. For a planned release containing features and/or bug fixes the MAJOR version will be incremented. 
+ 
+With the increased release cadence in order to allow proper planning for feature deprecation, Salt will introduce a minimum 1-year, expected 3 major release cycle. If `neon` ships in January 2020, any new deprecations will not be removed until January 2021 at the earliest.
+
+Note that this only applies to deprecations within Salt’s control. If a 3rd party changes, Salt may release updates earlier.
+
+## Support matrix 
+
+To be able to focus on the stability and innovation in the Salt platform, we will be adopting the industry-standard approach of no longer supporting older releases. Salt will provide select support for serious bugs and CVEs for the most recent release. Minor bug fixes will be targeted in the next scheduled Salt release.
+
+
+## Alternatives
+[alternatives]: #alternatives
+
+The primary alternative is to keep doing the things that Salt has been doing, and try to make incremental changes. The costs (both opportunity and real) are no longer acceptable.
+
+Long and unpredictable release times, unstable branches, missing (or extra) fixes and features, and regressions are all problems with our existing workflow. Trying to fix these problems here and there will not make the huge positive impact for our community that we are trying to make.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+Hopefully all the questions have been answered in this SEP. Upgrading Salt should continue to work like it always has - our changes are focused on the external development process of Salt. If you feel like you have unanswered questions, please come ask them at the Salt Office Hours on October 1st, 2019, or find us in [#salt on IRC on Freenode](http://webchat.freenode.net/?channels=salt&uio=Mj10cnVlJjk9dHJ1ZSYxMD10cnVl83), the [SaltStack Community Slack](https://saltstackcommunity.herokuapp.com/), or the [Salt Users mailing list](https://groups.google.com/forum/#!forum/salt-users).
```

Comment:
I think the biggest unanswered question I have is around deprecations.

As I understand this proposal, majors are released on a 90 day schedule. If a need arises for a minor, say because of a critical bug fix, then one will be issued at some point during the 90 day cycle, though this isn't always the case, and ideally (?) is _never_ the case.

So, assuming an ideal cycle where there are four releases a year, it seems like deprecations could happen at _any_ of those major releases. (Please correct me if this is wrong.) 

So, I have to wonder if this is setting ourselves up for a situation wherein most, or even all, upgrades are between majors, and so, even to to take in minor bug fixes, users are exposed to deprecations and a resulting change in behavior. I think we all agree that avoiding this bind is desirable.

Could there be room in this single-branch strategy for a major-minor model to still exist? For example, what if we took the four releases a year and structured them this way?

Quarter | Type | Policy
-- | -- | --
Q1 | Minor | (breaking changes NOT allowed)
Q2 | Minor | (breaking changes NOT allowed)
Q3 | Minor | (breaking changes NOT allowed)
Q4 | *Major* | (breaking changes *ARE* allowed)

IMHO, this keeps most of the benefits you've described here but it also allows users more predictability when it comes to upgrades. WDYT?

Reply:
> With the increased release cadence in order to allow proper planning for feature deprecation, Salt will introduce a minimum 1-year, expected 3 major release cycle. If `neon` ships in January 2020, any new deprecations will not be removed until January 2021 at the earliest.

I mean, I guess the answer is - yes there *could* be breaking changes in any release (which is not different from the current major release process). But they won't be "surprise here is something that you've never heard of!" kinds of changes, just ones that have seen a *minimum* of 1-year of warnings calling out the deprecation.

I expect we would require longer lead times for more invasive breaking changes.

Thinking about the alternative that you propose where we have a once-per-year breaking change release, and *feels* icky to me.

Initially I thought that could make sense, but the problem with that is that if we have **one** breaking change release per year we literally have to break **everything** that we're going to break in that release.

Instead of making minor changes over time - many of which will likely affect a subset of our user base, we would be making *different* changes at **one time** that would likely affect our entire user base.

The other side effect there is that there would be some kind of campaigning closer to the deprecation release to do things that would break for people with less warning. If we simply state that *every* deprecation needs at least a year of warning, that puts everyone on a level ground, and there's no temptation to "just make an exception this one time".

That may be a small side effect, but it is still there.

----

Ultimately, I prefer the *potential* for breaking changes with each release - this spreads out the potential issues: if there's something that wasn't caught in testing or RC with a deprecation, the hope is that it will be a minor issue, and we'll be able to get it out in the patch release. Worst case it will be super gnarly to fix and we'll just have to roll back that change in the patch release. But either case should cause much less impact overall.

Compare that to the case where we may have several deprecations go wrong in a single release. All of the sudden the world is on fire for everyone, and the core team is running around trying to fix all of the things. *Ick*.

I prefer small changes over time, to sweeping changes all at once (but, maybe I'm in the minority 🤷‍♂).